src/AutoMapper/MapperConfiguration.cs

```diff
@@ -201,40 +193,41 @@ public TypeMap ResolveTypeMap(Type sourceType, Type destinationType)
 
         public TypeMap ResolveTypeMap(TypePair typePair)
         {
-            var typeMap = _typeMapPlanCache.GetOrAdd(typePair);
-            if(typeMap != null && typeMap.MapExpression == null && Configuration.CreateMissingTypeMaps)
-            {
-                lock(typeMap)
-                {
-                    typeMap.Seal(this);
-                }
-            }
+            var typeMap = _typeMapPlanCache.GetOrAdd(typePair);
+
+            if (typeMap == null)
+                return null;
+
+            lock (typeMap)

```

Comment:
Yes, but the lock was taken only in certain conditions. You deleted those checks. That's my point :)